### DISCUSSION TITLE
Re-use previous run

### DISCUSSION THREAD

**Discussion by skoudoro** [community/community#26908] [Actions]
<p>Hi,</p>
<p>I would like to call a function from a sh script on a different run. Both functions are in the same script (my_script.sh). However, I would like to have a different “step name” on my GitHub action.</p>
<p>How can I achieve that?</p>
<p>The code below generates an error:  <code>my_function_2: command not found</code>.  It seems I need to call again <code>source my_script.sh</code> which I do not want.</p>
<pre><code class="lang-auto">- name: Setup My script
  run: |
    source my_script.sh
    my_function_1
- name: Call function2
  run: |
    # my_function_2 is defined in my_script
    my_function_2
</code></pre>
<p>thank you for your feedback</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>Each step gets a fresh Bash instance, so all loaded functions are lost, and you’ll have to source the script file again. Though in my opinion it’d be nicer to handle the function selection with command line options (maybe a subcommand?) instead of sourcing the script and then calling the function.</p>
<p>If sourcing/running the script twice creates undesirable side effects that may be the real problem to address.</p>

**Comment by skoudoro** [comment]
<p>Thank you <a class="mention" href="/airtower-luna">@airtower-luna</a> for your feedback.</p>
<p>I will follow your advice and review the design of my script and Github action.</p>

**Comment by airtower-luna** [comment]
<p>Oh, one thing I forgot earlier: You can create groups <em>within</em> the output of one step, with the <a href="https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#using-workflow-commands-to-access-toolkit-functions"><code>group</code> and <code>endgroup</code> workflow commands</a>. Not sure if that’d fit your needs, but it’d work like this:</p>
<pre><code class="lang-yaml">    - run: |
        source my_script.sh
        echo "::group::Run function 1"
        my_function_1
        echo "::endgroup::"
        echo "::group::Run function 2"
        my_function_2
        echo "::endgroup::"
</code></pre>
<p>In the log view this will create a folding structure inside the step.</p>
